components/places/src/history_sync/plan.rs

```diff
@@ -0,0 +1,619 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use std::collections::HashSet;
+use std::time::{SystemTime, UNIX_EPOCH};
+
+use rusqlite::{Connection};
+use sql_support::ConnExt;
+use url::{Url};
+
+use super::{MAX_OUTGOING_PLACES, MAX_VISITS};
+use super::record::{HistoryRecord, HistoryRecordVisit, HistorySyncRecord};
+use types::{SyncGuid, Timestamp, VisitTransition};
+use storage::history_sync::{
+    apply_synced_deletion,
+    apply_synced_visits,
+    apply_synced_reconcilliation,
+    fetch_outgoing,
+    fetch_visits,
+    FetchedVisit,
+    FetchedVisitPage,
+    finish_outgoing,
+    OutgoingInfo,
+};
+
+use api::history::{can_add_url};
+use error::*;
+
+use sync15_adapter::{
+    IncomingChangeset,
+    OutgoingChangeset,
+    Payload,
+};
+
+// In desktop sync, bookmarks are clamped to Jan 23, 1993 (which is 727747200000)
+// There's no good reason history records could be older than that, so we do
+// the same here (even though desktop's history currently doesn't)
+// XXX - there's probably a case to be made for this being, say, 5 years ago -
+// then all requests earlier than that are collapsed into a single visit at
+// this timestamp.
+const EARLIEST_TIMESTAMP: Timestamp = Timestamp(727747200000);
+
+/// Clamps a history visit date between the current date and the earliest
+/// sensible date.
+fn clamp_visit_date(visit_date: Timestamp) -> Timestamp {
+    let now = Timestamp::now();
+    if visit_date > now {
+        return visit_date;
+    }
+    if visit_date < EARLIEST_TIMESTAMP {
+        return EARLIEST_TIMESTAMP;
+    }
+    return visit_date;
+}
+
+#[derive(Debug)]
+pub enum IncomingPlan {
+    Skip, // An entry we just want to ignore - either due to the URL etc, or because no changes.
+    Invalid(Error), // Something's wrong with this entry.
+    Failed(Error), // The entry appears sane, but there was some error.
+    Delete(), // We should delete this.
+    // We should apply this. If SyncGuid is Some, then it is the existing guid
+    // for the same URL, and if that doesn't match the incoming record, we
+    // should change the existing item to the incoming one and upload a tombstone.
+    Apply(Option<SyncGuid>, Url, Option<String>, Vec<HistoryRecordVisit>),
+    Reconciled, // Entry exists locally and it's the same as the incoming record.
+}
+
+
+fn plan_incoming_record(conn: &Connection, record: HistoryRecord, max_visits: usize) -> IncomingPlan {
```

Comment:
Should we check that record has a sane places guid here? Once a generic guid type is available (#268) it probably will help here, but here is basically the code i'm using in my WIP/backlog branch for that issue, which we can just put in places in the mean time: https://gist.github.com/thomcc/e2af31d88d79bfcd4f5dd8eebc8f327d.